### PR TITLE
Time - Corinna

### PR DESCRIPTION
**Pull request by corinna-fab**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | In the pizza model I wrote a pizza_by_crust method to filter the pizzas and also helps the top_ten method.
Describe how you approached testing that model method. What edge cases did you come up with? | Testing to make sure the order is properly descending by vote count and looking at what happens when there are either no pizzas or no votes.
What are `session` and `flash`? What is the difference between them? | Both are hash-like objects but session holds information until the user closes their browsers, clears their cookies, etc. whereas flash holds information for two action cycles and is most often used to display messages (such as error messages when submitted a form).
What was one thing that you gained more clarity on through this assignment? | Migrations and how fickle and MEAN they can be. Most importantly, the value of a db:migrate.
What is the Heroku URL of your deployed application? | https://hidden-stream-11227.herokuapp.com/

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: app/views/pizzas/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      <% end %>
+    </tbody>                     
+    </table>
+  </section>
+<section class="media-table">
+  <h4>Thick Pizza</h4>
+  <table class="table">                         
+    <thead>
+      <tr>   
+        <th width="80">Votes</th>
```
Comment (beccaelenzil) [inline app/views/pizzas/index.html.erb:45]: You have the same code to show a list of works repeated 3 times. Could you use a view partial or a loop to DRY this up?

File: app/controllers/pizzas_controller.rb [hunk 2]
```diff
@@ -0,0 +1,117 @@
+class PizzasController < ApplicationController
+  # before_action :require_login, only: [:upvote]
+
+  def index
+    @pizzas = Pizza.paginate(:page=>params[:page],:per_page=>15).order(updated_at: :desc)
+  end
+
+  def show
+    pizza_id = params[:id]
```
Comment (beccaelenzil) [inline app/controllers/pizzas_controller.rb:9]: consider adding a controller filter `find_pizza` to DRY up your code.

File: app/controllers/pizzas_controller.rb [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    else
+      new_upvote = Vote.new(user_id: @user.id, pizza_id: @pizza.id)
+
+      if new_upvote.save
+        flash[:success] = "You have successfully voted on #{@pizza.name}!"
+        if @pizza.vote_count.nil?
+          @pizza.vote_count = 1
+        else
+          @pizza.vote_count += 1
+        end
```
Comment (beccaelenzil) [inline app/controllers/pizzas_controller.rb:86]: Consider moving these few lines of logic into a custom model method.

**Review by beccaelenzil — COMMENTED** [review]
Great job overall! Your implementation matches the demo site very closely, and the learning goals for this assignment were definitely met. I've left a few in-line comments for you to review. I love that you made it your own. Yummmm pizza. Keep up the hard work!

**Comment by beccaelenzil** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️There are not 10, but that's ok,
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️404
| 10. Verify that the "View all users" page lists no users | ✔️It list the users that are in the seeds
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️Consider using partials for homepage / all media view.
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | See in-line comment
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅
